text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,356 @@
+trait and type erasure / unnecessary boxing of errors to enable downcasting to
+extract members.
+[[1]](https://docs.rs/tracing-error/0.1.2/src/tracing_error/error.rs.html#269-274).
+
+## Use an alternative to Any for passing generic types across the trait boundary
+
+Nika Layzell has proposed an alternative implementation using a `Provider` type
+which avoids using `&dyn Any`. I do not necessarily think that the main
+suggestion is necessarily better, but it is much simpler.
+
+* https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0af9dbf0cd20fa0bea6cff16a419916b
+* https://github.com/mystor/object-provider
+
+With this design an implementation of the `provide_context` fn might instead look like:
+
+```rust
+fn provide_context<'r, 'a>(&'a self, request: Request<'r, 'a>) -> ProvideResult<'r, 'a> {
+    request
+        .provide::<PathBuf>(&self.path)?
+        .provide::<Path>(&self.path)?
+        .provide::<dyn Debug>(&self.path)
+}
+```
+
+The advantages of this design are that:
+
+1. It supports accessing trait objects and slices
+2. If the user specifies the type they are trying to pass in explicitly they
+   will get compiler errors when the type doesn't match.
+3. Takes advantage of deref sugar to help with conversions from wrapper types
+   to inner types.
+4. Less verbose implementation
+
+The disadvatages are:
+
+1. More verbose function signature, very lifetime heavy
+2. The Request type uses unsafe code which needs to be verified
+3. could encourage implementations where they pass the provider to
+   `source.provide` first which would prevent the error reporter from knowing
+   which error in the chain gathered each piece of context and might cause
+   context to show up multiple times in a report.
+
+# Prior art
+[prior-art]: #prior-art
+
+I do not know of any other languages whose error handling has similar
+facilities for accessing members when reporting errors. For the most part prior
+art exists within rust itself in the form of previous additions to the `Error`
+trait.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What should the names of these functions be?
+    - `context`/`context_ref`/`provide_context`
+    - `member`/`member_ref`
+    - `provide`/`request`
+- Should we go with the implementation that uses `Any` or the one that supports
+  accessing dynamically sized types like traits and slices?
+- Should there be a by value version for accessing temporaries?
+    - I bring this up specifically for the case where you want to use this
+      function to get an `Option<&[&dyn Error]>` out of an error, in this case
+      its unlikely that the error behind the trait object is actually storing
+      the errors as `dyn Errors`, and theres no easy way to allocate storage to
+      store the trait objects.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+I'd love to see the various error creating libraries like `thiserror` adding
```

Comment:
Prefer "the author" or "we" to "I" in something that'll become a normative text, IMO